### DISCUSSION TITLE
unhashable type for custom class.

### DISCUSSION THREAD

**Discussion by Hyeoghwan** [pydantic/pydantic#5159] [Question] [answered]
Here is my Code: 

```
from pydantic import BaseModel, Field, validator

class Image(BaseModel):
    width: int


class InputDto(BaseModel):
    images: List[Image] = Field(default_factory=list)

    @validator("images")
    def validate_images(cls, value: List[Image]) -> List[Image]:
        return list(set(value))


test = InputDto(images=[{'width': 1}, {'width': 2}])

```

and the code triggers the error: 

> pydantic.error_wrappers.ValidationError: 1 validation error for InputDto
images
  unhashable type: 'Image' (type=type_error)
  
  But if change the code liek this. it works. 

> @validator("images", each_item=True)

I don't know why..

**Comment by dmontagu** [comment, marked as answer]
You can make your `Image` instances hashable — which is necessary to put them into a set — by either 1, freezing them (though this would mean any `Image` with the same attributes was considered equal, not sure if that's what you want or not), or 2, adding the object.__hash__ function (then only exact instance matches will be equal).

Option 1:
```python
class Image(BaseModel):
    width: int

    class Config:
        frozen = True
```

Option 2:
```python
class Image(BaseModel):
    width: int

    __hash__ = object.__hash__
```

**Reply by Hyeoghwan** [reply]
Thanks! But why each_item makes the code work? just a curious..

**Reply by dmontagu** [reply]
It works with each_item because then the call to `set(Image(width=100))` causes the `Image.__iter__()` method to be called, which loops over tuples of the key-value pairs defining the model. In particular, `Image(width=100).__iter__()` yields `('width', 100)`, and so `list(set(Image(width=100)))` ends up being `[('width', 100)]`
